rfcs/2021-06-01-7694-improve-monitor-build-performance.md

```diff
@@ -0,0 +1,115 @@
+
+As many of Vector’s dependencies don’t often change, developers could provide benefit to themselves and others by “priming the pump” when they are the ones updating dependencies for chore PRs, and so on.  Their initial compilation would then benefit those down the line.
+
+Further, in the case of CI, systems like GitHub Actions are not contextually aware of the Rust build system.  While they may be able to naively cache outputs that land on disk, they are not aware of things like compiler flags, or compiler versions, and so could lead to confusing mismatch issues that waste time and effort to debug.
+
+### Rework “release” profile for non-versioned releases
+
+We currently utilize the same Cargo “release” profile whether we’re minting a versioned release or simply running a release build locally for performance testing.  While a boon for our users, our usage of LTO and modified “codegen-units” represents a significant increase in compilation time.  On a standard-issue laptop used by Vector engineers, it can take upwards of 40 minutes to build a release binary with these settings enabled.  With the settings disabled, the same build takes around 27 minutes, or 35% less time.
+
+These settings can be added back during the build process, when building versioned releases, without much effort.
+
+### Add system telemetry to Github Actions runners
+
+As Github Actions natively provides no telemetry of any kind about runners, or even basic metrics (how long did a job sit before running? etc), we’re often in the dark when it comes to CI and runner performance as a whole.
+
+We should be running the Datadog agent on every runner possible, collecting both high-level system metrics as well as more fine-grained metrics, such as Docker metrics on a per-container basis.  These metrics will, I believe, prove invaluable to debugging issues with the overall CI pipeline and its performance over time.  While understanding the build performance will likely come down to more advanced tooling used directly by engineers, we can’t optimize things like “use all the cores available” or “we’re using too much memory during CI runs” unless we actually have some data, any data, from those runs.
+
+### Manually instrument CI to report build performance
+
+Once we were properly utilizing Datadog on our CI runners, we could then instrument CI builds
+directly to report the build time to Datadog, allowing us to track the build performance over time.
+
+This data would be sparse given how many builds are likely to be triggered on a given day, but would
+form the start of tracking build performance holistically.
+
+## Rationale
+
+Currently, as mentioned above, it can take nearly 40 minutes for a clean release build of Vector.
+Depending on the hardware used, this can be reduced to 25 - 30 minutes, but the time spent is
+untenable even at the 25 - 30 minute range.  This build time not only permeates the local
```

Comment:
I am curious what the difference will be after these optimizations? I ask because it is worth considering better machines for our team. Is that a point we should add to this document?

Reply:
I think we can successfully avoid needing to purchase newer, more powerful hardware based on the optimizations we have available to us.

While I'm not sure we can apply these optimizations equally across the board, given the matrix of OS types, CPU architecture types, etc, used by developers... I am sure that for cases where they can all be applied together and used, their benefit will be an order of magnitude reduction in build time, if not more.  

Upgraded hardware likely represents a much smaller improvement on top of that, even if we think of utilizing high core count/high core frequency machines.